### PR TITLE
feat: PrevRealm ignores user realms in MsgRun

### LINKED ISSUE #1664: feat: `PrevRealm` has the same behavior between `MsgRun` and `MsgCall`
Currently, if `PrevRealm` is called directly in the context of `MsgRun`, it returns the *user realm* instead of the user address as in `MsgCall`.

### PR DESCRIPTION
**Pull request by tbruyelle**
Fix #1664

As commented in the IsRealm() method, a better format of user realms should emerge in the form of `gno.land/u/user_address`, which would remove the confusion between standard realms and the realm forged under the MsgRun transaction.

BREAKING CHANGE: `std.PrevRealm` is not returning the user realm any more when invoked under a transaction broadcasted by `MsgRun`.

To run the txtar test:
```
$ go test ./gno.land/cmd/gnoland/ -v -run TestTestdata/prevrealm
```

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: gnovm/pkg/gnolang/values.go [hunk 1]
```diff
@@ -816,8 +816,15 @@ type PackageValue struct {
...
+// IsRealm returns true if pv represents a realm.
+// A user realm, forged by the MsgRun transaction, isn't considered as a realm.
+// XXX The user realm format is:
+// PkgPath="gno.land/r/user_address" PkgName="main"
+// A less confusing format could emerge in the future, with the form of
+// PkgPath="gno.land/u/user_address"
 func (pv *PackageValue) IsRealm() bool {
-	return IsRealmPath(pv.PkgPath)
+	return IsRealmPath(pv.PkgPath) &&
+		pv.PkgName != "main" // Discard user realms forged by MsgRun
```
Comment (thehowl) [inline gnovm/pkg/gnolang/values.go:827]: packages can actually be called `main` (there is no restriction on upload) so this creates an attack vector (a realm could call another, while having IsRealm() == false). Can we check the PkgPath to match the format in vmkeeper (which is reserved) instead?

Comment (thehowl) [reply gnovm/pkg/gnolang/values.go:827]: Mhh, we could add `/u` straight away, but it's not necessary.

`sdk/vm/keeper.go`:

```
var reRunPath = regexp.MustCompile(`gno\.land/r/g[a-z0-9]+/run`)

// AddPackage adds a package with given fileset.
func (vm *VMKeeper) AddPackage(ctx sdk.Context, msg MsgAddPackage) error {
...
	if reRunPath.MatchString(pkgPath) {
		return ErrInvalidPkgPath("reserved package name: " + pkgPath)
	}
```

so we can use reRunPath in the opposite way (= if it matches, it's not a realm).

Though `/u/` sounds good to, you can re-purpose the PR to do that.

Comment (tbruyelle) [reply gnovm/pkg/gnolang/values.go]: Thank you for the details, indeed I can just this regexp, this is the way to go I think. The `/u/` change needs maybe more discussions, I'll leave it for later.
See f103d97

**Comment by leohhhn** [comment]
@tbruyelle If I understand correctly, this only modifies the PrevRealm in case of MsgRun? If so, I changed the name of the PR to make it more clear

**Comment by tbruyelle** [comment]
> @tbruyelle If I understand correctly, this only modifies the PrevRealm in case of MsgRun? If so, I changed the name of the PR to make it more clear

Yes, this is exactly that, thanks for the clarification.

**Review by zivkovicmilos — APPROVED** [review]

File: gno.land/cmd/gnoland/testdata/prevrealm.txtar [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+gnokey maketx addpkg -pkgdir $WORK/r/myrlm -pkgpath gno.land/r/myrlm -gas-fee 1000000ugnot -gas-wanted 2100000 -broadcast -chainid=tendermint_test test1
+stdout 'OK!'
+
+## deploy r/foo
+gnokey maketx addpkg -pkgdir $WORK/r/foo -pkgpath gno.land/r/foo -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+stdout 'OK!'
+
+## deploy p/demo/bar
+gnokey maketx addpkg -pkgdir $WORK/p/demo/bar -pkgpath gno.land/p/demo/bar -gas-fee 1000000ugnot -gas-wanted 2000000 -broadcast -chainid=tendermint_test test1
+stdout 'OK!'
```
Comment (thehowl) [inline gno.land/cmd/gnoland/testdata/prevrealm.txtar:36]: Use loadpkg?

File: gnovm/pkg/gnolang/realm.go [hunk 3]
```diff
@@ -1517,10 +1518,14 @@ func isUnsaved(oo Object) bool {
 // be realms and as such to have their state persisted. This is used by [IsRealmPath].
 const realmPathPrefix = "gno.land/r/"
 
+var ReGnoRunPath = regexp.MustCompile(`gno\.land/r/g[a-z0-9]+/run`)
```
Comment (thehowl) [inline gnovm/pkg/gnolang/realm.go:1521]: We should change this to /u/ maybe, but we'll leave that For Another Time(tm)...

**Review by thehowl — APPROVED** [review]
Looks mostly good.

File: gno.land/cmd/gnoland/testdata/prevrealm.txtar (hunk 2 above)
Comment (tbruyelle) [reply gno.land/cmd/gnoland/testdata/prevrealm.txtar]: done here 89f6938acbf9daac1c9614f5c1e627c772649cfd

**Pull request merged**